tough/src/editor/targets.rs

```diff
@@ -0,0 +1,536 @@
+    ) -> Result<&mut Self> {
+        let delegations = self.delegations.as_mut().context(error::NoDelegations)?;
+        let mut keyids = Vec::new();
+        for (keyid, key) in keys {
+            // Check to see if the key is present
+            if delegations
+                .keys
+                .iter()
+                .find(|(_, candidate_key)| key.clone().eq(candidate_key))
+                .is_none()
+            {
+                // Key isn't present yet, so we need to add it
+                delegations.keys.insert(keyid.clone(), key);
+            };
+            keyids.push(keyid.clone());
+        }
+
+        // If a role was provided add keyids to the delegated role
+        if let Some(role) = role {
+            for delegated_role in &mut delegations.roles {
+                if delegated_role.name == role {
+                    delegated_role.keyids.extend(keyids.clone());
+                }
+            }
+            for delegated_role in self.new_roles.get_or_insert(Vec::new()).iter_mut() {
+                if delegated_role.name == role {
+                    delegated_role.keyids.extend(keyids.clone());
+                }
+            }
+        }
+        Ok(self)
+    }
+
+    /// Removes a key from delegations keyids, if a role is specified only removes it from the role
+    pub fn remove_key(&mut self, keyid: &Decoded<Hex>, role: Option<&str>) -> Result<&mut Self> {
+        let delegations = self.delegations.as_mut().context(error::NoDelegations)?;
+        // If a role was provided remove keyid from the delegated role
+        if let Some(role) = role {
+            for delegated_role in &mut delegations.roles {
+                if delegated_role.name == role {
+                    delegated_role.keyids.retain(|key| keyid != key);
+                }
+            }
+        } else {
+            delegations.keys.remove(keyid);
+        }
+        Ok(self)
+    }
+
+    /// Delegates `paths` to `targets` adding a `DelegatedRole` to `new_roles`
+    pub fn delegate_role(
+        &mut self,
+        targets: Signed<DelegatedTargets>,
+        paths: PathSet,
+        key_pairs: HashMap<Decoded<Hex>, Key>,
+        keyids: Vec<Decoded<Hex>>,
+        threshold: NonZeroU64,
+    ) -> Result<&mut Self> {
+        self.add_key(key_pairs, None)?;
+        self.new_roles
+            .get_or_insert(Vec::new())
+            .push(DelegatedRole {
+                name: targets.signed.name,
+                paths,
+                keyids,
+                threshold,
+                terminating: false,
+                targets: Some(Signed {
+                    signed: targets.signed.targets,
+                    signatures: targets.signatures,
+                }),
+            });
+        Ok(self)
+    }
+
+    /// Removes a role from delegations if `recursive` is `false`
+    /// requires the role to be an immediate delegated role
+    /// if `true` removes whichever role eventually delegates 'role'
```

Comment:
There's a bit of run-on sentence going on here. Could we clarify the `recursive` bit? If recursive is "false", it means that we will only remove roles that this `targets` has delegated? We will not remove roles further down the chain right?